### PR TITLE
Magic Refactor + Wizard Grimoire

### PR DESCRIPTION
**Pull request by keronshb**
## About the PR

- [x] Split Spellbook and Magic Systems & Comps
- [x] Move Spellbook and Magic to Shared & network
- [x] Fix Fireball & Projectile Spells
- [x] Fix Sounds on some spells
- [x] Make a Speech Event so all speech is handled through it
- [x] Add conditions to spells that need robes, hat, staff, etc.
- [x] Refactor/rename targeting
- [x] Add a Master Spellbook & Wizard Currency via the shop system (bind to mind for main user)
- [x] Add a way to refund purchases (abilities and items)
- [x] Add a way to only buy/refund in a certain area
- [x] Add Action Upgrade purchases to Master Spellbook

_Will do in a different PR_
~~BONUS:~~
~~RPG Mechanic ECS for targeting and such (MMO-like floor AOEs, etc)~~
~~Field ECS to have a field do something (spawn a temp dance floor, time stop, time dilation, healing, etc)~~

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

Alpha example video of the Wizard's grimoire:

https://github.com/space-wizards/space-station-14/assets/54602815/bd7bafda-9a94-4c37-9afa-966b1da13d70

## Breaking changes

**Changelog**

### CODE REVIEW

**Comment by JoeHammad1844** [comment]
hold alt and type 0153 on the numpad for the ™ character

**Comment by JoeHammad1844** [comment]
mute toxin will be able to block wizards from casting, right?

**Comment by keronshb** [comment]
> mute toxin will be able to block wizards from casting, right?

Depends on the spell.

**Comment by Aisu9** [comment]
yus, all spells doesn't need a vocal component

File: Content.Shared/Actions/ActionUpgradeSystem.cs [hunk 1]
```diff
@@ -68,48 +69,75 @@ public bool TryUpgradeAction(EntityUid? actionId, ActionUpgradeComponent? action
...
+        return canLevel;
+    }
+
+    private bool CanUpgrade(int newLevel, Dictionary<int, EntProtoId> levelDict,  [NotNullWhen(true)]out EntProtoId? newLevelProto)
+    {
+        var canUpgrade = false;
+        newLevelProto = null;
+
+        var finalLevel = levelDict.Keys.ToList()[levelDict.Keys.Count - 1];
+
```
Comment (metalgearsloth) [inline Content.Shared/Actions/ActionUpgradeSystem.cs:119]: Dictionary ordering isn't guaranteed.

Comment (keronshb) [reply Content.Shared/Actions/ActionUpgradeSystem.cs:119]: This actually got merged anyway in #22277

**Comment by UbaserB** [comment]
Can you call the "master spellbook" a grimoire instead?

**Comment by K-Dynamic** [comment]
inb4 chemists throw muting foam grenades at wizard

**Comment by keronshb** [comment]
> Can you call the "master spellbook" a grimoire instead?

It's been renamed!

**Comment by SpaceCowboyServer** [comment]
Wanna ask, what do you think of placing each spell of SharedMagicSystem inside it's own file, like SharedMagicSystem.Knock.cs, so it would be a little bit easier to find particular spell among the files and file itself wouldn't become too large? 
I did this for wizard spells in my fork: https://github.com/space-wizards/space-station-14/blob/921a37b8e5bf217c569fa3b4710fc5d1a9ae6298/Content.Shared/Magic/SharedMagicSystem.cs

Of cause it's PR of yours and you're in charge here, but maybe you'd find this idea useful.

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Well i just pushed some minor cleanup and tried to implement the buying condition you mentioned to the catalog with ListingLimitedUpgradesCondition, it's broken cuz the prototype ID check in line 34 fails after upgrading once (maybe compare parents and have a common one between all upgrades? idk), still commited it in case you want to finish it, feel free to reverse that.

Things that can be done when the antag event is actually gonna be added (tho i recommend fixing these now):
- You can create ForceWall many times on a tile
- If you use a spell, refund it and then buy it again you don't have to wait for the cooldown to be over
-  Blink doesnt get deselected after a use
- Wands should prob have infinite ammo since you can refill them with the refund trick (unless you can fix it another way)
- Refunding items should require you to have them in your hands, failing to refund should have user feedback
- Is it intended that summon ghosts has no price? It seems pretty disruptive to not cost anything

Things that you need to do in this pr:
- Split the spell events with regions (partials may be a bit overkill) 

You can self merge if you want after that

**Comment by keronshb** [comment]
> Well i just pushed some minor cleanup and tried to implement the buying condition you mentioned to the catalog with ListingLimitedUpgradesCondition, it's broken cuz the prototype ID check in line 34 fails after upgrading once (maybe compare parents and have a common one between all upgrades? idk), still commited it in case you want to finish it, feel free to reverse that.
> 
> Things that can be done when the antag event is actually gonna be added (tho i recommend fixing these now):
> 
>     * You can create ForceWall many times on a tile
> 
>     * If you use a spell, refund it and then buy it again you don't have to wait for the cooldown to be over
> 
>     * Blink doesnt get deselected after a use
> 
>     * Wands should prob have infinite ammo since you can refill them with the refund trick (unless you can fix it another way)
> 
>     * Refunding items should require you to have them in your hands, failing to refund should have user feedback
> 
>     * Is it intended that summon ghosts has no price? It seems pretty disruptive to not cost anything
> 
> 
> Things that you need to do in this pr:
> 
>     * Split the spell events with regions (partials may be a bit overkill)
> 
> 
> You can self merge if you want after that

So refunding works only on the map they were purchased on.

It's intended that the refund button gets disabled once the wizard teleports from his ship to the station. So wherever they spawn it'll automatically mark the map id they're on and when they leave that map, the refund button won't work anymore.

I think the prototypeID thing broke because it went from ProductUpgradeID to ProductUpgradeId. Did you adjust spellbook_catalog to match?

**Comment by AJCM-git** [comment]
>I think the prototypeID thing broke because it went from ProductUpgradeID to ProductUpgradeId. Did you adjust spellbook_catalog to match?

Yeah, that one i did., tho I forgot to fix another one, next commit will fix the linter.
Also, what i meant is that I implemented the condition mentioned in the TODO but it's not working cuz it compares prototype ids  when it should be doing something else (see ListingLimitedUpgradesCondition line 34), still pushed it on the last commit in case you wanted to fix that. Feel free to reverse if my implementation is completely wrong

**Comment by keronshb** [comment]
> > I think the prototypeID thing broke because it went from ProductUpgradeID to ProductUpgradeId. Did you adjust spellbook_catalog to match?
> 
> Yeah, that one i did., tho I forgot to fix another one, should be fixed now. Also, what i meant is that I implemented the condition mentioned in the TODO but it's not working cuz it compares prototype ids when it should be doing something else (see ListingLimitedUpgradesCondition line 34), still pushed it 2 commits ago in case you wanted to fix that. Feel free to reverse if my implementation is completely wrong

I found the issue, it's checking by .Count instead of the highest level. The system's allowed to skip levels so the count wouldn't match up.

So you could have two effected levels of level 2 and level 4, which would still have a Count of 2 and it would exit early.

**Comment by keronshb** [comment]
> Well i just pushed some minor cleanup and tried to implement the buying condition you mentioned to the catalog with ListingLimitedUpgradesCondition, it's broken cuz the prototype ID check in line 34 fails after upgrading once (maybe compare parents and have a common one between all upgrades? idk), still commited it in case you want to finish it, feel free to reverse that.
> 
> Things that can be done when the antag event is actually gonna be added (tho i recommend fixing these now):
> 
>     * You can create ForceWall many times on a tile
> 
>     * If you use a spell, refund it and then buy it again you don't have to wait for the cooldown to be over
> 
>     * Blink doesnt get deselected after a use
> 
>     * Wands should prob have infinite ammo since you can refill them with the refund trick (unless you can fix it another way)
> 
>     * Refunding items should require you to have them in your hands, failing to refund should have user feedback
> 
>     * Is it intended that summon ghosts has no price? It seems pretty disruptive to not cost anything
> 
> 
> Things that you need to do in this pr:
> 
>     * Split the spell events with regions (partials may be a bit overkill)
> 
> 
> You can self merge if you want after that

Force wall was lowered to 12 seconds for despawn time. Generally the spell should be ready to cast before the wall disappears but now at least it won't let you spam it at a huge net positive.

Summon ghosts is mimicking ss13 as a free show ghosts to all. I haven't figured out how to get it to continually show new ghosts so I allowed it to be recast after a few minutes.

**Review by AJCM-git — APPROVED** [review]

**Pull request merged**